File: ai/shared/languages.py
"""
Backend-side counterpart to frontend/src/lib/languages.js.

Listing a language here means the UI is allowed to *offer* it as a
choice, not that Bhashini can currently translate into it — pipeline
support depends on the configured Bhashini models/pipeline and can
change independently of this static list. The backend stays
authoritative for actual capability: GET /api/v1/languages reports both
this catalog and whether a real translation backend is configured at
all (see ai.translation.translator_configured), never a promise that
every pair here works.

Kept in one place, in this shape, specifically so it never has to be
hand-copied in sync with the frontend file — a future iteration could
serve the frontend list from this endpoint directly instead of
duplicating it, but until then the two are kept in lockstep by hand and
this docstring is the reminder to do so.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class LanguageEntry:
    code: str  # None-equivalent is "en" here; the API always has a real code
    name: str
    native_name: str


LANGUAGE_CATALOG: tuple[LanguageEntry, ...] = (
    LanguageEntry("en", "English", "English"),
    LanguageEntry("hi", "Hindi", "हिन्दी"),
    LanguageEntry("bn", "Bengali", "বাংলা"),
    LanguageEntry("te", "Telugu", "తెలుగు"),
    LanguageEntry("mr", "Marathi", "मराठी"),
    LanguageEntry("ta", "Tamil", "தமிழ்"),
    LanguageEntry("ur", "Urdu", "اردو"),
    LanguageEntry("gu", "Gujarati", "ગુજરાતી"),
    LanguageEntry("kn", "Kannada", "ಕನ್ನಡ"),
    LanguageEntry("or", "Odia", "ଓଡ଼ିଆ"),
    LanguageEntry("ml", "Malayalam", "മലയാളം"),
    LanguageEntry("pa", "Punjabi", "ਪੰਜਾਬੀ"),
    LanguageEntry("as", "Assamese", "অসমীয়া"),
    LanguageEntry("mai", "Maithili", "मैथिली"),
    LanguageEntry("sat", "Santali", "ᱥᱟᱱᱛᱟᱲᱤ"),
    LanguageEntry("ks", "Kashmiri", "كٲشُر"),
    LanguageEntry("ne", "Nepali", "नेपाली"),
    LanguageEntry("sd", "Sindhi", "سنڌي"),
    LanguageEntry("kok", "Konkani", "कोंकणी"),
    LanguageEntry("doi", "Dogri", "डोगरी"),
    LanguageEntry("mni", "Manipuri", "মৈতৈলোন্"),
    LanguageEntry("bo", "Bodo", "བོད་སྐད"),
    LanguageEntry("sa", "Sanskrit", "संस्कृतम्"),
)

File: frontend/src/lib/languages.js
/**
 * Single source of truth for the languages the UI can offer as an answer
 * language. Previously this list (in lib/api.js) only had English, Hindi
 * and Kannada, backing three header buttons. It now covers English plus
 * all 22 languages scheduled in the Eighth Schedule to the Constitution,
 * which is what a "Language ▾" dropdown needs to be worth having.
 *
 * Listing a language here means the UI can *offer* it — it does not mean
 * Bhashini can actually translate into it right now. Pair capability
 * depends on the configured pipeline/model and can change independently
 * of this file, so the backend remains authoritative: see
 * GET /api/v1/languages (backend/app/api/routes.py), which returns the
 * catalog it actually knows how to serve plus whether a real translation
 * backend is configured at all. This file is the fallback shown before
 * that call resolves, and the shape both agree on.
 */

export const LANGUAGE_CATALOG = [
  { code: null, bhashiniCode: 'en', label: 'English', englishName: 'English', nativeName: 'English' },
  { code: 'hi', bhashiniCode: 'hi', label: 'हिन्दी', englishName: 'Hindi', nativeName: 'हिन्दी' },
  { code: 'bn', bhashiniCode: 'bn', label: 'বাংলা', englishName: 'Bengali', nativeName: 'বাংলা' },
  { code: 'te', bhashiniCode: 'te', label: 'తెలుగు', englishName: 'Telugu', nativeName: 'తెలుగు' },
  { code: 'mr', bhashiniCode: 'mr', label: 'मराठी', englishName: 'Marathi', nativeName: 'मराठी' },
  { code: 'ta', bhashiniCode: 'ta', label: 'தமிழ்', englishName: 'Tamil', nativeName: 'தமிழ்' },
  { code: 'ur', bhashiniCode: 'ur', label: 'اردو', englishName: 'Urdu', nativeName: 'اردو' },
  { code: 'gu', bhashiniCode: 'gu', label: 'ગુજરાતી', englishName: 'Gujarati', nativeName: 'ગુજરાતી' },
  { code: 'kn', bhashiniCode: 'kn', label: 'ಕನ್ನಡ', englishName: 'Kannada', nativeName: 'ಕನ್ನಡ' },
  { code: 'or', bhashiniCode: 'or', label: 'ଓଡ଼ିଆ', englishName: 'Odia', nativeName: 'ଓଡ଼ିଆ' },
  { code: 'ml', bhashiniCode: 'ml', label: 'മലയാളം', englishName: 'Malayalam', nativeName: 'മലയാളം' },
  { code: 'pa', bhashiniCode: 'pa', label: 'ਪੰਜਾਬੀ', englishName: 'Punjabi', nativeName: 'ਪੰਜਾਬੀ' },
  { code: 'as', bhashiniCode: 'as', label: 'অসমীয়া', englishName: 'Assamese', nativeName: 'অসমীয়া' },
  { code: 'mai', bhashiniCode: 'mai', label: 'मैथिली', englishName: 'Maithili', nativeName: 'मैथिली' },
  { code: 'sat', bhashiniCode: 'sat', label: 'ᱥᱟᱱᱛᱟᱲᱤ', englishName: 'Santali', nativeName: 'ᱥᱟᱱᱛᱟᱲᱤ' },
  { code: 'ks', bhashiniCode: 'ks', label: 'كٲشُر', englishName: 'Kashmiri', nativeName: 'كٲشُر' },
  { code: 'ne', bhashiniCode: 'ne', label: 'नेपाली', englishName: 'Nepali', nativeName: 'नेपाली' },
  { code: 'sd', bhashiniCode: 'sd', label: 'سنڌي', englishName: 'Sindhi', nativeName: 'سنڌي' },
  { code: 'kok', bhashiniCode: 'kok', label: 'कोंकणी', englishName: 'Konkani', nativeName: 'कोंकणी' },
  { code: 'doi', bhashiniCode: 'doi', label: 'डोगरी', englishName: 'Dogri', nativeName: 'डोगरी' },
  { code: 'mni', bhashiniCode: 'mni', label: 'মৈতৈলোন্', englishName: 'Manipuri', nativeName: 'মৈতৈলোন্' },
  { code: 'bo', bhashiniCode: 'bo', label: 'བོད་སྐད', englishName: 'Bodo', nativeName: 'བོད་སྐད' },
  { code: 'sa', bhashiniCode: 'sa', label: 'संस्कृतम्', englishName: 'Sanskrit', nativeName: 'संस्कृतम्' },
];

/** Look up a catalog entry by the value the app uses internally (the
 * `code` field — `null` for English, matching LangCtx's existing
 * "null means detect/English" convention). */
export function findLanguage(code) {
  return LANGUAGE_CATALOG.find(l => l.code === code) ?? LANGUAGE_CATALOG[0];
}
